File: src/python/flows_specification/publisher.py
"""Secure publication of the accepted specification bundle."""

from __future__ import annotations

from datetime import datetime, timezone
import hashlib
import json
from pathlib import Path
import shutil
import tempfile

from harness_engine import atomic_io, docs_reader

from . import renderer, store

DESTINATION = Path("specs/active")
MANIFEST = store.DIR / "publish-manifest.json"
EXPECTED_FILES = (
    "00-prd.md",
    "10-software-requirements-specification.md",
    "20-software-design-document.md",
    "30-readiness-handoff.md",
)
DEVELOPMENT_PLAN_FILENAME = "40-development-plan.json"
EXPECTED_PUBLISHED_FILES = EXPECTED_FILES + (DEVELOPMENT_PLAN_FILENAME,)


def _digest(content: str) -> str:
    return "sha256:" + hashlib.sha256(content.encode("utf-8")).hexdigest()


def render_all(prd: dict, srs: dict, sdd: dict, readiness: dict) -> dict[str, str]:
    """Render the exact four documents that ``publish`` will write."""
    return {
        EXPECTED_FILES[0]: renderer.prd(prd),
        EXPECTED_FILES[1]: renderer.srs(srs),
        EXPECTED_FILES[2]: renderer.sdd(sdd),
        EXPECTED_FILES[3]: renderer.readiness(readiness),
    }


def build_development_plan(srs: dict, sdd: dict, readiness: dict, rendered: dict[str, str]) -> dict:
    requirements = {
        item.get("id"): item.get("statement", "")
        for item in (srs.get("functionalRequirements", []) + srs.get("qualityRequirements", []))
        if item.get("id")
    }
    adrs = {item.get("id"): item for item in sdd.get("adrs", []) if item.get("id")}
    slice_ids = {item.get("id"): index + 1 for index, item in enumerate(readiness.get("slices", []))}
    features = []
    for index, item in enumerate(readiness.get("slices", [])):
        requirement_ids = set(item.get("requirementIds", []))

        def linked(artifact: dict) -> bool:
            return bool(requirement_ids.intersection(artifact.get("requirementIds", [])))

        acceptance_criteria = [artifact for artifact in srs.get("acceptanceCriteria", []) if linked(artifact)]
        interfaces = [artifact for artifact in srs.get("interfaces", []) if linked(artifact)]
        data_rules = [artifact for artifact in srs.get("dataRules", []) if linked(artifact)]
        controls = [artifact for artifact in sdd.get("controls", []) if linked(artifact)]
        requirement_text = [f"{ref}: {requirements[ref]}" if ref in requirements else ref for ref in item.get("requirementIds", [])]
        decisions = []
        for ref in item.get("adrIds", []):
            adr = adrs.get(ref)
            decisions.append(
                f"{ref}: {adr.get('title', '')}. Decision: {adr.get('decision', '')}. Rationale: {adr.get('rationale', '')}"
                if adr else ref
            )
        references = list(dict.fromkeys(
            item.get("requirementIds", [])
            + item.get("adrIds", [])
            + [artifact.get("id") for artifact in acceptance_criteria]
            + [artifact.get("id") for artifact in interfaces]
            + [artifact.get("id") for artifact in data_rules]
            + [artifact.get("id") for artifact in controls]
        ))
        acceptance_text = [item.get("acceptanceCriterion", "")] + [
            f"{artifact.get('id', '')}: Given {artifact.get('given', '')}. "
            f"When {artifact.get('when', '')}. Then {artifact.get('then', '')}"
            for artifact in acceptance_criteria
        ]
        related_contract_text = [
            f"{artifact.get('id', '')}: {artifact.get('name', '')}. {artifact.get('description', '')}"
            for artifact in interfaces + controls
        ] + [
            f"{artifact.get('id', '')}: {artifact.get('rule', '')}"
            for artifact in data_rules
        ]
        features.append({
            "id": index + 1,
            "title": item.get("goal", ""),
            "priority": index + 1,
            "passes": False,
            "dependsOn": [slice_ids[ref] for ref in item.get("dependsOn", []) if ref in slice_ids],
            "description": (
                f"{item.get('goal', '')} Observable outcome: {item.get('observableOutcome', '')}. "
                f"Happy path: {item.get('happyPath', '')}. Failure path: {item.get('failurePath', '')}."
            ),
            "references": references,
            "implementationContext": {
                "requirements": requirement_text,
                "decisions": decisions,
                "constraints": [f"out of scope: {value}" for value in item.get("outOfScope", [])]
                + item.get("contracts", [])
                + related_contract_text,
                "files": [item.get("suggestedTarget", "")],
                "acceptance": acceptance_text,
            },
        })
    digest_source = "|".join(rendered[name] for name in EXPECTED_FILES)
    return {
        "schema": "iao/development-plan/v1",
        "specificationBundleDigest": _digest(digest_source),
        "sourceFiles": list(EXPECTED_FILES),
        "features": features,
        "targetDescription": (srs.get("delivery") or {}).get("target", ""),
        "verificationDescription": (srs.get("delivery") or {}).get("verificationStrategy", ""),
    }


def _read_manifest() -> dict | None:
    try:
        return json.loads(MANIFEST.read_text())
    except (OSError, json.JSONDecodeError):
        return None


def _ownership_error() -> str | None:
    manifest = _read_manifest() or {}
    owned = set(manifest.get("ownedFiles", []))
    if not DESTINATION.exists():
        return None
    for existing in DESTINATION.iterdir():
        if not existing.is_file():
            continue
        if existing.name not in EXPECTED_PUBLISHED_FILES:
            return f"unrecognized file '{existing.name}' exists in '{DESTINATION}'; publish blocked."
        if existing.name not in owned:
            return f"file '{existing.name}' exists in '{DESTINATION}' but is not owned by a previous publish; publish blocked."
    return None


def verify_postcondition(expected_digests: dict[str, str]) -> tuple[bool, str | None]:
    content, files = docs_reader.read(str(DESTINATION))
    if files != list(EXPECTED_FILES):
        return (
            False,
            f"postcondition failed: DocsReader returned files {files}, expected {list(EXPECTED_FILES)}.",
        )
    for filename in EXPECTED_FILES:
        path = DESTINATION / filename
        try:
            text = path.read_text()
        except OSError as error:
            return False, f"postcondition failed: could not read '{path}': {error}"
        actual = _digest(text)
        if expected_digests.get(filename) != actual:
            return (
                False,
                f"postcondition failed: '{filename}' digest does not match the staged digest.",
            )
        if text.rstrip() not in content:
            return (
                False,
                f"postcondition failed: DocsReader content does not contain '{filename}' in full.",
            )
    return True, None


def publish(
    prd: dict, srs: dict, sdd: dict, readiness: dict
) -> tuple[bool, str | None]:
    rendered = render_all(prd, srs, sdd, readiness)
    rendered[DEVELOPMENT_PLAN_FILENAME] = json.dumps(
        build_development_plan(srs, sdd, readiness, rendered), separators=(",", ":"), ensure_ascii=False
    )
    ownership_error = _ownership_error()
    if ownership_error:
        return False, ownership_error

    staging = Path(tempfile.mkdtemp(prefix="specification-", dir="."))
    try:
        digests: dict[str, str] = {}
        for filename, content in rendered.items():
            (staging / filename).write_text(content)
            digests[filename] = _digest(content)

        DESTINATION.mkdir(parents=True, exist_ok=True)
        for filename in EXPECTED_PUBLISHED_FILES:
            atomic_io.write_text_atomic(
                str(DESTINATION / filename), (staging / filename).read_text()
            )

        manifest = {
            "ownedFiles": list(EXPECTED_PUBLISHED_FILES),
            "fileDigests": digests,
            "manifestDigest": _digest(
                "|".join(digests[name] for name in EXPECTED_PUBLISHED_FILES)
            ),
            "publishedAt": datetime.now(timezone.utc).isoformat(),
        }
        store.DIR.mkdir(parents=True, exist_ok=True)
        atomic_io.write_text_atomic(
            str(MANIFEST), json.dumps(manifest, separators=(",", ":"))
        )
        return verify_postcondition(digests)
    finally:
        shutil.rmtree(staging, ignore_errors=True)

File: src/rust/flows_specification/src/publisher.rs
//! Publisher (mirrors `SpecificationPublisher.cs`): renders, stages, validates and publishes
//! the four accepted documents to `specs/active/`.

use crate::renderer::render_all;
use crate::store::{digest_str, read, write};
use serde_json::{Value, json};
use std::{
    collections::{HashMap, HashSet},
    fs,
    path::Path,
};

pub(crate) const DEST_DIR: &str = "specs/active";
pub(crate) const FILENAMES: [&str; 4] = [
    "00-prd.md",
    "10-software-requirements-specification.md",
    "20-software-design-document.md",
    "30-readiness-handoff.md",
];
pub(crate) const DEVELOPMENT_PLAN_FILENAME: &str = "40-development-plan.json";

fn expected_published_filenames() -> Vec<String> {
    FILENAMES
        .iter()
        .map(|name| (*name).to_string())
        .chain(std::iter::once(DEVELOPMENT_PLAN_FILENAME.to_string()))
        .collect()
}

fn development_plan(
    srs: &Value,
    sdd: &Value,
    readiness: &Value,
    rendered: &HashMap<String, String>,
) -> Value {
    let mut requirements = HashMap::new();
    for item in srs["functionalRequirements"]
        .as_array()
        .into_iter()
        .flatten()
        .chain(srs["qualityRequirements"].as_array().into_iter().flatten())
    {
        if let (Some(id), Some(statement)) = (item["id"].as_str(), item["statement"].as_str()) {
            requirements.insert(id.to_string(), statement.to_string());
        }
    }
    let mut adrs = HashMap::new();
    for item in sdd["adrs"].as_array().into_iter().flatten() {
        if let Some(id) = item["id"].as_str() {
            adrs.insert(id.to_string(), item);
        }
    }
    let slices = readiness["slices"].as_array().cloned().unwrap_or_default();
    let slice_ids: HashMap<String, i32> = slices
        .iter()
        .enumerate()
        .filter_map(|(index, item)| {
            item["id"]
                .as_str()
                .map(|id| (id.to_string(), index as i32 + 1))
        })
        .collect();
    let features: Vec<Value> = slices.iter().enumerate().map(|(index, item)| {
        let requirement_ids = item["requirementIds"].as_array().cloned().unwrap_or_default();
        let linked = |artifact: &Value| artifact["requirementIds"].as_array().into_iter().flatten().any(|id| requirement_ids.iter().any(|required| required == id));
        let acceptance_criteria: Vec<Value> = srs["acceptanceCriteria"].as_array().into_iter().flatten().filter(|artifact| linked(artifact)).cloned().collect();
        let interfaces: Vec<Value> = srs["interfaces"].as_array().into_iter().flatten().filter(|artifact| linked(artifact)).cloned().collect();
        let data_rules: Vec<Value> = srs["dataRules"].as_array().into_iter().flatten().filter(|artifact| linked(artifact)).cloned().collect();
        let controls: Vec<Value> = sdd["controls"].as_array().into_iter().flatten().filter(|artifact| linked(artifact)).cloned().collect();
        let adr_ids = item["adrIds"].as_array().cloned().unwrap_or_default();
        let requirements_text: Vec<Value> = requirement_ids.iter().filter_map(|value| value.as_str()).map(|id| json!(requirements.get(id).map(|text| format!("{id}: {text}")).unwrap_or_else(|| id.to_string()))).collect();
        let decisions: Vec<Value> = adr_ids.iter().filter_map(|value| value.as_str()).map(|id| {
            if let Some(adr) = adrs.get(id) { json!(format!("{id}: {}. Decision: {}. Rationale: {}", adr["title"].as_str().unwrap_or(""), adr["decision"].as_str().unwrap_or(""), adr["rationale"].as_str().unwrap_or("")))} else { json!(id) }
        }).collect();
        let mut references: Vec<Value> = requirement_ids.iter().chain(adr_ids.iter()).cloned().collect();
        for artifact in acceptance_criteria.iter().chain(interfaces.iter()).chain(data_rules.iter()).chain(controls.iter()) {
            if let Some(id) = artifact["id"].as_str() { references.push(json!(id)); }
        }
        let depends_on: Vec<Value> = item["dependsOn"].as_array().into_iter().flatten().filter_map(|value| value.as_str().and_then(|id| slice_ids.get(id)).map(|id| json!(id))).collect();
        let out_of_scope: Vec<Value> = item["outOfScope"].as_array().into_iter().flatten().filter_map(|value| value.as_str()).map(|value| json!(format!("out of scope: {value}"))).chain(item["contracts"].as_array().into_iter().flatten().cloned()).collect();
        let mut related_contracts = Vec::new();
        for artifact in interfaces.iter().chain(controls.iter()) {
            related_contracts.push(json!(format!(
                "{}: {}. {}",
                artifact["id"].as_str().unwrap_or(""),
                artifact["name"].as_str().unwrap_or(""),
                artifact["description"].as_str().unwrap_or("")
            )));
        }
        for artifact in &data_rules {
            related_contracts.push(json!(format!(
                "{}: {}",
                artifact["id"].as_str().unwrap_or(""),
                artifact["rule"].as_str().unwrap_or("")
            )));
        }
        let mut acceptance_text = vec![json!(item["acceptanceCriterion"].as_str().unwrap_or(""))];
        for artifact in &acceptance_criteria {
            acceptance_text.push(json!(format!(
                "{}: Given {}. When {}. Then {}",
                artifact["id"].as_str().unwrap_or(""),
                artifact["given"].as_str().unwrap_or(""),
                artifact["when"].as_str().unwrap_or(""),
                artifact["then"].as_str().unwrap_or("")
            )));
        }
        let goal = item["goal"].as_str().unwrap_or("");
        json!({
            "id": index + 1, "title": goal, "priority": index + 1, "passes": false,
            "dependsOn": depends_on,
            "description": format!("{goal} Observable outcome: {}. Happy path: {}. Failure path: {}.", item["observableOutcome"].as_str().unwrap_or(""), item["happyPath"].as_str().unwrap_or(""), item["failurePath"].as_str().unwrap_or("")),
            "references": references,
            "implementationContext": {"requirements": requirements_text, "decisions": decisions, "constraints": out_of_scope.into_iter().chain(related_contracts).collect::<Vec<_>>(), "files": [item["suggestedTarget"].as_str().unwrap_or("")], "acceptance": acceptance_text}
        })
    }).collect();
    let joined = FILENAMES
        .iter()
        .map(|name| rendered.get(*name).cloned().unwrap_or_default())
        .collect::<Vec<_>>()
        .join("|");
    json!({"schema": "iao/development-plan/v1", "specificationBundleDigest": digest_str(&joined), "sourceFiles": FILENAMES, "features": features, "targetDescription": srs["delivery"]["target"].as_str().unwrap_or(""), "verificationDescription": srs["delivery"]["verificationStrategy"].as_str().unwrap_or("")})
}

fn random_hex() -> String {
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    format!("{:x}-{:x}", std::process::id(), nanos)
}

/// Independent postcondition check (blueprint 0004 §6 item 6): re-reads `specs/active/`
/// through the REAL `docs_reader::read` (never a re-implementation) and byte-compares its
/// output against what `publish()` just wrote — names/order, per-file digest, and that
/// DocsReader's concatenation contains each file's full on-disk text (detects truncation).
fn verify_postcondition(expected_digests: &HashMap<String, String>) -> Result<(), String> {
    let (content, files) = harness_engine::docs_reader::read(DEST_DIR);
    let expected: Vec<String> = FILENAMES.iter().map(|s| s.to_string()).collect();
    if files != expected {
        return Err(format!(
            "postcondition failed: DocsReader.Read('{DEST_DIR}') returned files [{}], expected [{}].",
            files.join(", "),
            expected.join(", ")
        ));
    }
    for name in FILENAMES {
        let file_path = format!("{DEST_DIR}/{name}");
        let text = fs::read_to_string(&file_path)
            .map_err(|e| format!("postcondition failed: could not read '{file_path}': {e}"))?;
        let actual_digest = digest_str(&text);
        let expected_digest = expected_digests.get(name).cloned().unwrap_or_default();
        if actual_digest != expected_digest {
            return Err(format!(
                "postcondition failed: '{name}' on-disk digest '{actual_digest}' does not match the digest recorded at publish time '{expected_digest}'."
            ));
        }
        if !content.contains(text.trim_end()) {
            return Err(format!(
                "postcondition failed: DocsReader.Read('{DEST_DIR}')'s content does not contain the full on-disk text of '{name}' (possible truncation)."
            ));
        }
    }
    Ok(())
}

/// Renders, stages, validates and — only if validation passes — publishes the four accepted
/// documents. Never deletes/globs `specs/active/`: every write is a named copy of one of the
/// four known filenames. The staging directory is always cleaned up (`finally`-equivalent via
/// the closure below), and a failed ownership check leaves the destination untouched.
pub(crate) fn publish(
    prd: &Value,
    srs: &Value,
    sdd: &Value,
    readiness: &Value,
) -> Result<HashMap<String, String>, String> {
    let mut rendered = render_all(prd, srs, sdd, readiness);
    let plan = serde_json::to_string(&development_plan(srs, sdd, readiness, &rendered))
        .map_err(|e| format!("failed to serialize development plan: {e}"))?;
    rendered.insert(DEVELOPMENT_PLAN_FILENAME.to_string(), plan);
    let expected_published = expected_published_filenames();
    let staging_dir = format!("{DEST_DIR}.staging-{}", random_hex());

    let result = (|| -> Result<HashMap<String, String>, String> {
        fs::create_dir_all(&staging_dir)
            .map_err(|e| format!("failed to create staging dir: {e}"))?;

        let mut digests = HashMap::new();
        for name in &expected_published {
            let content = rendered.get(name).cloned().unwrap_or_default();
            let staged_path = format!("{staging_dir}/{name}");
            harness_engine::atomic_io::write_atomic(Path::new(&staged_path), &content)
                .map_err(|e| format!("failed to stage '{name}': {e}"))?;
            digests.insert(name.to_string(), digest_str(&content));
        }

        // Ownership check BEFORE anything in the destination is touched.
        let previously_owned: HashSet<String> = read("publish-manifest.json")
            .and_then(|manifest| manifest["ownedFiles"].as_array().cloned())
            .unwrap_or_default()
            .into_iter()
            .filter_map(|item| item.as_str().map(String::from))
            .collect();
        let expected: HashSet<&str> = expected_published.iter().map(String::as_str).collect();
        if let Ok(entries) = fs::read_dir(DEST_DIR) {
            for entry in entries.flatten() {
                if entry.file_type().map(|t| t.is_file()).unwrap_or(false) {
                    let name = entry.file_name().to_string_lossy().to_string();
                    if !expected.contains(name.as_str()) {
                        return Err(format!(
                            "unrecognized file '{name}' exists in '{DEST_DIR}'; publish blocked."
                        ));
                    }
                    if !previously_owned.contains(&name) {
                        return Err(format!(
                            "file '{name}' exists in '{DEST_DIR}' but is not owned by a previous publish of this flow; publish blocked."
                        ));
                    }
                }
            }
        }

        // Validation passed: copy the four known files in, then write the manifest last.
        fs::create_dir_all(DEST_DIR).map_err(|e| format!("failed to create '{DEST_DIR}': {e}"))?;
        for name in &expected_published {
            let staged_path = format!("{staging_dir}/{name}");
            let dest_path = format!("{DEST_DIR}/{name}");
            fs::copy(&staged_path, &dest_path)
                .map_err(|e| format!("failed to publish '{name}': {e}"))?;
        }

        let manifest_digest = digest_str(
            &expected_published
                .iter()
                .map(|n| digests.get(n).cloned().unwrap_or_default())
                .collect::<Vec<_>>()
                .join("|"),
        );
        let manifest = json!({
            "ownedFiles": expected_published,
            "fileDigests": digests,
            "manifestDigest": manifest_digest,
            "publishedAt": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Micros, false),
        });
        write("publish-manifest.json", &manifest);

        verify_postcondition(&digests)?;

        Ok(digests)
    })();

    let _ = fs::remove_dir_all(&staging_dir);

    result
}
